metricbeat/module/kubernetes/util/kubernetes.go

```diff
@@ -129,155 +149,402 @@ func getResource(resourceName string) kubernetes.Resource {
 	}
 }
 
-// NewResourceMetadataEnricher returns an Enricher configured for kubernetes resource events
+// getExtraWatchers returns a list of the extra resources to watch based on some resource.
+// The full list can be seen in https://github.com/elastic/beats/issues/37243, at Expected Watchers section.
+func getExtraWatchers(resourceName string, config *kubernetesConfig) []string {
+	switch resourceName {
+	case PodResource:
+		extra := []string{NamespaceResource, NodeResource}
+		// We need to create watchers for ReplicaSets and Jobs that it might belong to,
+		// in order to be able to retrieve 2nd layer Owner metadata like in case of:
+		// Deployment -> Replicaset -> Pod
+		// CronJob -> job -> Pod
+		if config.AddResourceMetadata != nil && config.AddResourceMetadata.Deployment {
+			extra = append(extra, ReplicaSetResource)
+		}
+		if config.AddResourceMetadata != nil && config.AddResourceMetadata.CronJob {
+			extra = append(extra, JobResource)
+		}
+		return extra
+	case ServiceResource:
+		return []string{NamespaceResource}
+	case DeploymentResource:
+		return []string{NamespaceResource}
+	case ReplicaSetResource:
+		return []string{NamespaceResource}
+	case StatefulSetResource:
+		return []string{NamespaceResource}
+	case DaemonSetResource:
+		return []string{NamespaceResource}
+	case JobResource:
+		return []string{NamespaceResource}
+	case CronJobResource:
+		return []string{NamespaceResource}
+	case PersistentVolumeResource:
+		return []string{}
+	case PersistentVolumeClaimResource:
+		return []string{NamespaceResource}
+	case StorageClassResource:
+		return []string{}
+	case NodeResource:
+		return []string{}
+	case NamespaceResource:
+		return []string{}
+	default:
+		return []string{}
+	}
+}
+
+// getWatchOptions builds the kubernetes.WatchOptions{} needed for the watcher based on the config and nodeScope
+func getWatchOptions(config *kubernetesConfig, nodeScope bool, client k8sclient.Interface, log *logp.Logger) (*kubernetes.WatchOptions, error) {
+	var err error
+	options := kubernetes.WatchOptions{
```

Comment:
If by default Namespace filter  is part of Options, this means that resources that are not namespace aware will have this filter option. Like namespace and node. I don't know if that causes any error but we could avoid adding it for namespace and node if possible.

Reply:
Ok, how should we go about it then? Should there be a new function similar to `getResource` and `getExtraWatchers` that goes over every resource and returns if they are namespace and node aware? I am separating this new function from `getExtraWatchers` since we have yet to decide which resources start which watchers

Reply:
A simple approach could be:
1. in `getWatchOptions` we return just the default options without the namespace filter.
```
func getWatchOptions(config *kubernetesConfig, nodeScope bool, client k8sclient.Interface, log *logp.Logger) (*kubernetes.WatchOptions, error) {
	var err error
	defaultOptions := kubernetes.WatchOptions{
		SyncTimeout: config.SyncPeriod,
	}

	// Watch objects in the node only
	if nodeScope {
		nd := &kubernetes.DiscoverKubernetesNodeParams{
			ConfigHost:  config.Node,
			Client:      client,
			IsInCluster: kubernetes.IsInCluster(config.KubeConfig),
			HostUtils:   &kubernetes.DefaultDiscoveryUtils{},
		}
		defaultOptions.Node, err = kubernetes.DiscoverKubernetesNode(log, nd)
		if err != nil {
			return nil, fmt.Errorf("couldn't discover kubernetes node: %w", err)
		}
	}
	return &defaultOptions, err
}
```
2.  Then inside `createWatcher` we add the namespace filter to the defaultOptions except when the resource name is one of `NamespaceResource`, `NodeResource`, `PersistentVolumeResource`, `StorageClassResource`. All the rest seem to be namespaced. 

What do you think?

Reply:
Ok, does Namespace resource 

> Then inside createWatcher we add the namespace filter to the defaultOptions except when the resource name is one of NamespaceResource,

does not have namespace in the watch options?

Also, I added this change:

https://github.com/elastic/beats/blob/4cc8a47ca891b64c66a50ddb0ed6149c7e0e5d6e/metricbeat/module/kubernetes/util/kubernetes.go#L364-L375

We take care of namespace resource as being namespace aware, right?

Reply:
Maybe the approach for namespace aware is not the best using extra watchers, but defining instead a condition with the resource name

Reply:
The namespace resource has already a namespacewatcher and every time an event is triggered the namespace meta are updated and added to the new event. It does not need the namespacemeta in the `NewNamespaceAwareResourceMetadataGenerator`. 

But I don't think the namespaceAware part in `createMetadataGen` in needed at all. We can always use the `NewNamespaceAwareResourceMetadataGenerator`. 
As far as I can see from the code https://github.com/elastic/elastic-agent-autodiscover/blob/b42be0df903aba6580b80761af6df86fa80c845d/kubernetes/metadata/resource.go#L100

Then before the 
https://github.com/elastic/elastic-agent-autodiscover/blob/b42be0df903aba6580b80761af6df86fa80c845d/kubernetes/metadata/resource.go#L136 
 is invoked to add the namespace meta  the code checks if the `accessor.GetNamespace()` exists. In case it is a resource that is not namespace aware (like namespace or node) this will be empty so actually the code block which would add again the namespace metadata is not called.

Reply:
You are right about the options, thank you for the links. I refactored the whole part: the options.namespace and `createMetadataGen`

Reply:
Yes, I misunderstood.

I ended up reverting that change then. But the condition is still there, just to check if the namespace watcher is nil or not

Reply:
Yes you are right. This check is needed